### PR TITLE
Curvature Estimation Example 

### PR DESCRIPTION
**Pull request by sneha-belkhale**
Hey all, 

Here is the example that we talked about in #13074 . 

I compute a curvature estimation for the heart geometry, and display the raw values as a shader. I also added another shader that uses the curvature attribute to drive the specular highlights and a noise function for texture.

Let me know your thoughts / suggestions.

Best,
Sneha

### CODE REVIEW

**Comment by Mugen87** [comment]
For testing:

https://rawgit.com/sneha-belkhale/three.js/dev/examples/webgl_curvature_estimation.html

**Comment by Mugen87** [comment]
Is it necessary to add so a big model? It has round about 18 MB.

If you can't use one of the [existing models](https://github.com/mrdoob/three.js/tree/dev/examples/models) in the repo, i suggest you decrease the size of your file to < 5 MB or something like that.

**Comment by WestLangley** [comment]
Also, what feature of three.js this is demonstrating?

Check the console warnings.

File: examples/webgl_curvature_estimation.html [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+
+				requestAnimationFrame( animate );
+
+				render();
+
+			}
+
+			function render() {
+
+				controls.update();
```
Comment (Mugen87) [inline examples/webgl_curvature_estimation.html]: Not necessary to update the controls in this case...

File: examples/webgl_curvature_estimation.html [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+								dict[ strC ][ strB ] = c2b;
+
+							}
+
+							curvatureDict = {};
+							var min = 10, max = 0;
+
+							Object.keys( dict ).forEach( function( key ) {
+
+    						curvatureDict[ key ] = average( dict[ key ] );
```
Comment (Mugen87) [inline examples/webgl_curvature_estimation.html]: There are some minor formatting issues like this indentation here...

**Review by Mugen87 — CHANGES_REQUESTED** [review]

**Comment by sneha-belkhale** [comment]
@Mugen I checked out the existing models, and I think the ninja model will work. I have updated the model in the latest commit.

@WestLangley Fixed the console warnings. The example demonstrates how to compute and use custom geometry attributes ( curvature in this case ). Other attributes can be calculated and used to drive interesting shader effects.

**Comment by mrdoob** [comment]
Can you remove `heart.obj` from the PR?

**Comment by sneha-belkhale** [comment]
@mrdoob yes, it's removed now

File: examples/webgl_curvature_estimation.html [hunk 3]
```diff
@@ -0,0 +1,530 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>three.js webgl - shader - curvature [ninja]</title>
+		<div class="label" style="position: absolute;left: 20%;bottom: 10%;">Raw curvature values</div>
+		<div class="label" style="position: absolute;left: 68%;bottom: 10%;">Curvature driving specular + texture </div>
```
Comment (mrdoob) [inline examples/webgl_curvature_estimation.html]: These `<div>`s should go inside `<body>`.

File: examples/webgl_curvature_estimation.html [hunk 4]
```diff
@@ -0,0 +1,530 @@
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>three.js webgl - shader - curvature [ninja]</title>
+		<div class="label" style="position: absolute;left: 20%;bottom: 10%;">Raw curvature values</div>
```
Comment (mrdoob) [inline examples/webgl_curvature_estimation.html]: Hmm, what are these `<div>`s for?

Comment (sneha-belkhale) [reply examples/webgl_curvature_estimation.html]: they are the descriptions under the two scenes   ("Raw curvature values" and "Curvature driving specular + texture")

**Comment by WestLangley** [comment]
Why are you setting all the `needsUpdate` flags? They should not be required.

What is `cnoise()` for?

>Also, what feature of three.js this is demonstrating?
>>The example demonstrates how to compute and use custom geometry attributes ( curvature in this case )

I think this is best suited as a showcase, not as an example. IMO, as written, the javascript and shader code are too complicated to be useful to users as an example.

**Comment by sneha-belkhale** [comment]
> Why are you setting all the needsUpdate flags? They should not be required.

When switching filter modes (convex vs. concave) I need to update the curvature attribute to feed in the correct values to the shader. 

> What is cnoise() for?

cnoise is for generating a noise pattern based on vertex point position, that is combined with the curvature attribute in order to get a final texture.

> I think this is best suited as a showcase, not as an example. IMO, as written, the javascript and shader code are too complicated to be useful to users as an example.

I think this would be cool as an example, because it shows how to iterate over faces in a buffer geometry and calculate a meaningful attribute based on vertex positions and normals. The javascript may seem complicated, but using the bufferGeometry is innately complicated ... The shader code is actually very simple but looks complicated because of the cnoise() function, which I can remove if you all think would be helpful

File: examples/webgl_curvature_estimation.html (hunk 4 above)
Comment (sneha-belkhale) [reply examples/webgl_curvature_estimation.html]: Oh, I'll move them to <body>

**Comment by WestLangley** [comment]
How about removing the menu, using just your simple "raw" shader, and showing both positive and negative curvature in the same render?

Something like so, but modified according to your preference:
```glsl
gl_FragColor = vec4( 2. * vCurvature + 0.25, 0.0, - 2. * vCurvature + 0.25, 1.0 );
```
screen shot 2018-01-18 at 6 42 52 pm

**Comment by sneha-belkhale** [comment]
> How about removing the menu, using just your simple "raw" shader, and showing both positive and negative curvature in the same render?

I agree we could just use the raw shader for simplicity, but I think that showing both curvatures in the same render is hard to visually parse. 
I think the menu is nice to have and inline with the other three.js examples. Alternatively, maybe we can have two renders ( one for concave and one for convex )?

**Comment by WestLangley** [comment]
I will bow out of this thread, as it is up to you and @mrdoob. I was just expressing my opinion that this is better suited as a Showcase due to its complexity.

>I agree we could just use the raw shader for simplicity

Yes, please do that, if this will be used as an example. Thanks.

**Comment by sneha-belkhale** [comment]
@WestLangley Okay, appreciate your help and I will definitely simplify the example.

@mrdoob Let me know what you think about having just the raw shader with a menu to toggle between concave and convex curvature, and I can update the example

**Comment by mrdoob** [comment]
I agree with @WestLangley. I would just have the raw shader. I would keep the dropdown though.

**Comment by bhouston** [comment]
BTW here is C# source code for principle curvature (https://en.wikipedia.org/wiki/Principal_curvature) directions on a tri-mesh - it is useful for a lot of mesh optimization situations.  This is my old C# code that was ported from the original C++ (which I give credit to in the comments):

https://gist.github.com/bhouston/0be120012c6c9286de60aa5e596fcf14

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**